src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -450,80 +454,222 @@ private bool TryCreateCountSegment(string segmentText)
 
         /// <summary>
-        /// Tries to handle the segment as $ref. If it is $ref, then the rest of the path will be parsed/validated in this call.
+        /// Creates a filter clause from a navigation source and filter expression.
         /// </summary>
-        /// <param name="text">The text of the segment.</param>
-        /// <returns>Whether the text was $ref.</returns>
-        private bool TryCreateEntityReferenceSegment(string text)
+        /// <param name="navigationSource">Navigation source to which the filter refers.</param>
+        /// <param name="targetEdmType">Target type for the entity being referenced.</param>
+        /// <param name="filter">Filter expression.</param>
+        /// <returns>Filter clause with the navigation source and filter information.</returns>
+        private FilterClause GenerateFilterClause(IEdmNavigationSource navigationSource, IEdmType targetEdmType, string filter)
         {
-            string identifier;
-            string parenthesisExpression;
-            ExtractSegmentIdentifierAndParenthesisExpression(text, out identifier, out parenthesisExpression);
+            Debug.Assert(navigationSource != null, "navigationSource != null");
+            Debug.Assert(targetEdmType != null, "targetEdmType != null");
+            Debug.Assert(filter.Length > 0, "filter.Length > 0");
 
-            if (!this.IdentifierIs(UriQueryConstants.RefSegment, identifier))
+            ODataPathInfo currentOdataPathInfo = new ODataPathInfo(targetEdmType, navigationSource);
+
+            // Get the syntactic representation of the filter expression.
+            UriQueryExpressionParser expressionParser = new UriQueryExpressionParser(
+                configuration.Settings.FilterLimit, configuration.EnableCaseInsensitiveUriFunctionIdentifier);
+            QueryToken filterToken = expressionParser.ParseFilter(filter);
+
+            // Bind it to metadata.
+            BindingState state = new BindingState(configuration, currentOdataPathInfo.Segments.ToList())
+            {
+                ImplicitRangeVariable = NodeFactory.CreateImplicitRangeVariable(
+                currentOdataPathInfo.TargetEdmType.ToTypeReference(), currentOdataPathInfo.TargetNavigationSource)
+            };
+            state.RangeVariables.Push(state.ImplicitRangeVariable);
+
+            MetadataBinder binder = new MetadataBinder(state);
+            FilterBinder filterBinder = new FilterBinder(binder.Bind, state);
+
+            return filterBinder.BindFilter(filterToken);
+        }
+
+        /// <summary>
+        /// Try to handle the segment as $filter.
+        /// </summary>
+        /// <param name="segmentText">The raw segment text.</param>
+        /// <returns>Whether the segment was $filter.</returns>
+        /// <remarks>$filter path segment is different from existing path segments in that it strictly
+        /// follows the format of "$filter(@a)", where @a represents a parameter alias. Thus, this function

```

Comment:
>@a [](start = 43, length = 2)

should now be $filter(expression) where expression could be an alias or inline expression.

Reply:
Thanks for the catch!